aten/src/ATen/native/transformers/cuda/sdp_utils.h

```diff
@@ -70,13 +86,14 @@ inline std::array<SDPBackend, num_backends> priority_order(sdp_params params) {
     bool more_threads_cutlass = (threads_cutlass / 2) >= threads_flash;
     bool small_threads_flash = threads_flash < 60;
     bool large_head_dim = head_dim.max(params.key.sym_size(3)) == 128;
-    if ((small_threads_flash && more_threads_cutlass) || large_head_dim) {
-      return {
-          SDPBackend::efficient_attention,
-          SDPBackend::flash_attention,
-          SDPBackend::math};
+
+    // The training heuristic is taken from https://github.com/pytorch/pytorch/pull/99644
+    // Revisit when updated cutlass kernel is upstreamed.
+    if (check_requires_grad(params, false)) {
+      if (6 * threads_flash > query_lengths) return efficient_first;
+    } else if ((small_threads_flash && more_threads_cutlass) || large_head_dim)
```

Comment:
does the 94% accuracy use this combined heuristic or just `(6 * threads_flash > query_lengths)`?

Reply:
The 94% accuracy is just with the 6 * threads_flash > query_lengths. I was only gathering data for fw + bw together (which is why there's a requires_grad guard).

@danthe3rd mentioned that the heuristic in xformers is only for the forwards pass, and in xformers they can use say, cutlass fw + flash bw (which I don't think we're able to do).